### PR TITLE
Adapt the predict function, add scaling

### PR DESCRIPTION
**Pull request by Sakaaaa**
In the main.py, a call to the function of scaling is added (if the patch size is not equal 40X40)

The input paramaters of predict functions are changed, so that you can run it easily by setting the 'lTrain' = False in cnn_main.py. (But I only make this in network motion_new_CNN2D) There should be further changes in predict function, because I don't know how to choose a best training model, and use loops to predict using all models.

And about the train_after_rezise.py : now it's used to train and predict directly, if the patched data already exists.

### CODE REVIEW

File: config/paramnormal.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Datasets
```
Comment (thomaskuestner) [inline config/paramnormal.yml]: Don't push redundant config file -> delete this one

File: main.py [hunk 2]
```diff
@@ -9,12 +9,10 @@
...
 import cnn_main
+import scaling
 
-# parse parameters
-# training or prediction
-lTrain = True
-lSave = False # save intermediate test, training sets
-with open('config' + os.sep + 'param.yml', 'r') as ymlfile:
+
+with open('config' + os.sep + 'paramnormal.yml', 'r') as ymlfile:
```
Comment (thomaskuestner) [inline main.py]: leave old config loading in place

File: main.py [hunk 3]
```diff
@@ -81,16 +79,31 @@
     # perform splitting
     X_train, y_train, X_test, y_test = ttsplit.fSplitDataset(dAllPatches, dAllLabels, dAllPats, cfg['sSplitting'], patchSize, cfg['patchOverlap'], cfg['dSplitval'], '')
 
+    if patchSize[0] != 40:
```
Comment (thomaskuestner) [inline main.py]: Why only != 40 ?
Patch size is parametrizable -> what does this function perform? Up- and downscaling of the patches prior to training?
-> If yes, should be done inside the patching routine not in main

File: main.py [hunk 4]
```diff
@@ -81,16 +79,31 @@
...
-        hf.create_dataset('y_test', data=y_test)
-        hf.create_dataset('patchSize', data=patchSize)
-        hf.create_dataset('patchOverlap', data=cfg['patchOverlap'])
+         hf.create_dataset('X_train', data=X_train)
+         hf.create_dataset('X_test', data=X_test)
+         hf.create_dataset('y_train', data=y_train)
+         hf.create_dataset('y_test', data=y_test)
+         hf.create_dataset('patchSize', data=patchSize)
+         hf.create_dataset('patchOverlap', data=cfg['patchOverlap'])
+         hf.create_dataset('OrigPatchSize', data=OrigPatchSize)
```
Comment (thomaskuestner) [inline main.py]: How does patchsize look like resp. where do you define the scaled patchsizes?
I would suggest to use the config patchSize parameter as follows:
patchSize = [x1, y1, z1, x2, y2, z2, ...]
where x1 corresponds to 1st patching along x direction, y1: 1st patching along y direction z1: 1st patching along z direction, so e.g. for the 2D case with 2 patchSizes: patchSize = [40, 40, , 20, 20, ]
=> makes OrigPatchSize obsolete

File: main.py [hunk 5]
```diff
@@ -81,16 +79,31 @@
...
 
+# parse parameters
+# training or prediction
+lTrain = True
+lSave = False  # save intermediate test, training sets
 # perform training
-for iFold in range(0, len(X_train)):
-    cnn_main.fRunCNN({'X_train': X_train[iFold], 'y_train': y_train[iFold], 'X_test': X_test[iFold], 'y_test': y_test[iFold], 'patchSize': patchSize}, cfg['network'], lTrain, cfg['sOpti'], sOutPath, cfg['batchSize'], cfg['lr'], cfg['epochs'])
+for iFold in range(0,2):
```
Comment (thomaskuestner) [inline main.py]: why only range(0,2) ? 
-> makes no sense

File: main.py [hunk 6]
```diff
@@ -81,16 +79,31 @@
...
+#for iFold in range(0,len(X_train)):
+    if cfg['sSplitting'] == 'crossvalidation_patient':
+        CV_Patient = iFold + 1
+    cnn_main.fRunCNN({'X_train': X_train[iFold], 'y_train': y_train[iFold], 'X_test': X_test[iFold], 'y_test': y_test[iFold], 'patchSize': patchSize}, cfg['network'], lTrain, cfg['sOpti'], sOutPath, cfg['batchSize'], cfg['lr'], cfg['epochs'], CV_Patient)
```
Comment (thomaskuestner) [inline main.py]: hand over iFold or iFold+1 instead of creating new variable

File: networks/motion/CNN2D/2D_CNN.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
+import os.path
```
Comment (thomaskuestner) [inline networks/motion/CNN2D/2D_CNN.py]: deprecated file 2D_CNN -> rename to CNN2D
Is this your own file or does this resemble e.g. motion_all_CNN2D?

File: networks/motion/CNN2D/motion_all_CNN2D.py [hunk 8]
```diff
@@ -128,6 +128,15 @@ def createModel(patchSize, architecture='new'):
 
 def fTrain(X_train, y_train, X_test, y_test, sOutPath, patchSize, batchSizes=None, learningRates=None, iEpochs=None):
```
Comment (thomaskuestner) [inline networks/motion/CNN2D/motion_all_CNN2D.py]: this file is also on the old version!
please pull from master

File: networks/motion/CNN2D/motion_head_CNN2D.py [hunk 9]
```diff
@@ -28,15 +28,14 @@
...
-                            init='he_normal',
+                            (14, 14),
+                            kernel_initializer='he_normal',
                            # activation='sigmoid',
                             weights=None,
-                            border_mode='valid',
-                            subsample=(1, 1),
-                            W_regularizer=l2(1e-6),
-                            input_shape=(1, patchSize[0,0], patchSize[0,1])))
+                            padding='valid',
```
Comment (thomaskuestner) [inline networks/motion/CNN2D/motion_head_CNN2D.py]: changes not necessary
leave it as in master

File: networks/motion/CNN2D/motion_new_CNN2D.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
+import os.path
```
Comment (thomaskuestner) [inline networks/motion/CNN2D/motion_new_CNN2D.py]: deprecated file
-> see current master branch

File: networks/motion/CNN3D/3D_CNN.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
+import os.path
```
Comment (thomaskuestner) [inline networks/motion/CNN3D/3D_CNN.py]: deprecated file
Please pull first current master from branch and merge it local with your code, then push online and request a pull to master

File: train_after_resize.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+
+import h5py
+import yaml
+import os
+import cnn_main
+
+with open('config' + os.sep + 'paramnormal.yml', 'r') as ymlfile:
+    cfg = yaml.safe_load(ymlfile)
+
+forig = h5py.File('/home/s1241/no_backup/s1241/CNNArt/MultiScale/Headcross/8080/testout/crossVal8080to4040.h5', 'r')
```
Comment (thomaskuestner) [inline train_after_resize.py]: no hardcoded paths

File: train_after_resize.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (thomaskuestner) [inline train_after_resize.py]: Why is this file needed?
Can't we combine it with the main.py?

File: scaling.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
+# scale the 20X20 patches to 40X40 using cubic spline interpolation
+
+import h5py
+import numpy as np
+from scipy import interpolate
+import cnn_main
+import time
+
+def fscalling(X_train, X_test, patchSize):
+    afterSize = 40 # one value
```
Comment (thomaskuestner) [inline scaling.py]: no hard-coding -> use parameter as specified in config

File: scaling.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+    for ifold in range(len(X_train)):
+        lenTrain = X_train[ifold].shape[0]
+        lenTest = X_test[ifold].shape[0]
+        zaxis_train = np.arange(lenTrain)
+        zaxis_test = np.arange(lenTest)
+
+        ############ in batches
+        ########### Up scaling the training set
+        Batch=20
```
Comment (thomaskuestner) [inline scaling.py]: no hard-coding

File: scaling.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+            dAllx_test = dFoldx_test
+        else:
+            dAllx_test = np.concatenate((dAllx_test, dFoldx_test), axis=0)
+
+    patchSize = [afterSize, afterSize]
+    return dAllx_train, dAllx_test, patchSize
+
+if __name__ == "__main__":  ##### main for test directly
+    
+    forig = h5py.File('/home/s1241/no_backup/s1241/CNNArt/MultiScale/Headnormal/8080/testout/normal8080.h5','r')
```
Comment (thomaskuestner) [inline scaling.py]: no specific paths

File: scaling.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    y_train=forig['y_train']
+    y_test=forig['y_test']
+    patchSize=forig['patchSize']
+    patchOverlap=forig['patchOverlap']
+    
+    NX_train, NX_test, NewpatchSize = fscalling(X_train, X_test, patchSize)
+    OrigPatchSize = patchSize
+    patchSize = NewpatchSize
+    
+    with h5py.File('/home/s1241/no_backup/s1241/CNNArt/MultiScale/Headnormal/8080/testout/normal8080to4040.h5', 'w') as hf:
```
Comment (thomaskuestner) [inline scaling.py]: no specific paths

File: main.py (hunk 4 above)
Comment (Sakaaaa) [reply main.py]: The OrigiPatchSize was designed to creat the saving output path. e.g. the training results when using 4040 patches will be saved in path '~/headcross/4040/testout'. If the pateches are resized from 8080 to 4040, the result should be saved in '~/headcross/8080/testout'.

File: main.py (hunk 6 above)
Comment (Sakaaaa) [reply main.py]: But if the 'Splitting'= 'normal'
the 'CV_Patient' should be 0, 
the last input parameter can't be iFold+1

File: main.py [hunk 18]
```diff
@@ -94,16 +88,18 @@
          hf.create_dataset('y_test', data=y_test)
          hf.create_dataset('patchSize', data=patchSize)
          hf.create_dataset('patchOverlap', data=cfg['patchOverlap'])
-         hf.create_dataset('OrigPatchSize', data=OrigPatchSize)
+
+# Perform resizing of patches if needed
+X_train, X_test, NewpatchSize, sDatafile = scaling.fscalling(X_train, X_test, patchSize, sDatafile)
```
Comment (thomaskuestner) [inline main.py]: typo: fscalling (remove double "l")

File: scaling.py [hunk 19]
```diff
@@ -25,8 +28,8 @@ def fscalling(X_train, X_test, patchSize):
         zaxis_train = np.arange(lenTrain)
         zaxis_test = np.arange(lenTest)
 
-        ############ in batches
-        ########### Up scaling the training set
+        # in batches
+        # Up scaling the training set
         Batch=20
```
Comment (thomaskuestner) [inline scaling.py]: Why batch = 20?
Load it from config file

File: scaling.py [hunk 20]
```diff
@@ -90,7 +94,7 @@ def fscalling(X_train, X_test, patchSize):
     OrigPatchSize = patchSize
     patchSize = NewpatchSize
     
-    with h5py.File('/home/s1241/no_backup/s1241/CNNArt/MultiScale/Headnormal/8080/testout/normal8080to4040.h5', 'w') as hf:
+    with h5py.File('/home/s1241/no_backup/s1241/MultiScale/Headnormal/8080/testout/normal8080to4040.h5', 'w') as hf:
```
Comment (thomaskuestner) [inline scaling.py]: no path hard-coding

File: config/paramnormal.yml [hunk 21]
```diff
@@ -5,7 +5,7 @@ motion_head: &id001
    labelref  : [0]
```
Comment (thomaskuestner) [inline config/paramnormal.yml]: please do not push your own config file into master
include new parameters in the default param.yml file

Comment (thomaskuestner) [reply config/paramnormal.yml]: remove from commit!

File: main.py (hunk 18 above)
Comment (thomaskuestner) [reply main.py]: typo still existing: remove doulbe "l" in fscalling -> fscaling

File: scaling.py [hunk 22]
```diff
@@ -5,7 +5,14 @@
 import cnn_main
 import time
 
+def fscalling(X_train, X_test, patchSize, sDatafile):
```
Comment (thomaskuestner) [inline scaling.py]: typo: fscaling!

File: scaling.py (hunk 19 above)
Comment (thomaskuestner) [reply scaling.py]: still hard-coded batch size? load from config

File: cnn_main.py [hunk 23]
```diff
@@ -158,14 +158,14 @@ def fRunCNN(dData, sModelIn, lTrain, sParaOptim, sOutPath, iBatchSize, iLearning
...
+                                dData['patchSize'], iBatchSize, iLearningRate, iEpochs, CV_Patient)
 
         else:  # no optimization or grid search (if batchSize|learningRate are arrays)
             cnnModel.fTrain(dData['X_train'], dData['y_train'], dData['X_test'], dData['y_test'], sOutPath,
-                            dData['patchSize'], iBatchSize, iLearningRate, iEpochs)
+                            dData['patchSize'], iBatchSize, iLearningRate, iEpochs, CV_Patient)
 
     else:  # predicting
-        cnnModel.fPredict(dData['X_test'], dData['y_test'], dData['model_name'], sOutPath, dData['patchSize'], iBatchSize[0])
+        cnnModel.fPredict(dData['X_test'], dData['y_test'], sOutPath, dData['patchSize'], iBatchSize, iLearningRate, CV_Patient)
```
Comment (thomaskuestner) [inline cnn_main.py]: change to iBatchSize[0], iLearningRate[0]

File: config/paramnormal.yml (hunk 1 above)
Comment (thomaskuestner) [reply config/paramnormal.yml]: Remove!

File: main.py [hunk 24]
```diff
@@ -9,9 +9,10 @@
 import utils.DataPreprocessing as datapre
 import utils.Training_Test_Split as ttsplit
 import cnn_main
+import scaling
 
-# parse parameters
-with open('config' + os.sep + 'param.yml', 'r') as ymlfile:
+
+with open('config' + os.sep + 'paramnormal.yml', 'r') as ymlfile:
```
Comment (thomaskuestner) [inline main.py]: do not change here!

File: main.py [hunk 25]
```diff
@@ -84,6 +85,13 @@
         # perform splitting
         X_train, y_train, X_test, y_test = ttsplit.fSplitDataset(dAllPatches, dAllLabels, dAllPats, cfg['sSplitting'], patchSize, cfg['patchOverlap'], cfg['dSplitval'], '')
 
+<<<<<<< HEAD
```
Comment (thomaskuestner) [inline main.py]: why only:
`if patchSize[0] != 40`
better put a method naming, e.g. 
`if cfg['networkScaling'] == 'multiscale'`

File: main.py [hunk 26]
```diff
@@ -94,10 +102,36 @@
                 hf.create_dataset('y_test', data=y_test)
                 hf.create_dataset('patchSize', data=patchSize)
                 hf.create_dataset('patchOverlap', data=cfg['patchOverlap'])
+                hf.create_dataset('OrigPatchSize', data=OrigPatchSize)
 
     # perform training
     for iFold in range(0, len(X_train)):
         cnn_main.fRunCNN({'X_train': X_train[iFold], 'y_train': y_train[iFold], 'X_test': X_test[iFold], 'y_test': y_test[iFold], 'patchSize': patchSize}, cfg['network'], lTrain, cfg['sOpti'], sOutPath, cfg['batchSize'], cfg['lr'], cfg['epochs'])
+=======
```
Comment (thomaskuestner) [inline main.py]: you are training and afterwards retrain with the scaled patches?
If so: We should put in a separate "training method", see comment above for multiscale training.

File: networks/motion/CNN2D/2D_CNN.py (hunk 7 above)
Comment (thomaskuestner) [reply networks/motion/CNN2D/2D_CNN.py]: still deprecated, please rename!

File: networks/motion/CNN2D/motion_all_CNN2D.py [hunk 27]
```diff
@@ -128,6 +128,15 @@ def createModel(patchSize, architecture='new'):
...
     # grid search on batch_sizes and learning rates
+    # parse inputs
+    batchSizes = 64 if batchSizes is None else batchSizes
```
Comment (thomaskuestner) [inline networks/motion/CNN2D/motion_all_CNN2D.py]: that are still the old version. It should be:
`batchSize = [64] if batchSize is None else batchSize
learningRate = [0.01] if learningRate is None else learningRate
iEpochs = 300 if iEpochs is None else iEpochs`

File: networks/motion/CNN2D/motion_head_CNN2D.py (hunk 9 above)
Comment (thomaskuestner) [reply networks/motion/CNN2D/motion_head_CNN2D.py]: did you change them for compatibility or why?

File: networks/motion/CNN2D/motion_new_CNN2D.py (hunk 10 above)
Comment (thomaskuestner) [reply networks/motion/CNN2D/motion_new_CNN2D.py]: still deprecated

File: networks/motion/CNN3D/3D_CNN.py (hunk 11 above)
Comment (thomaskuestner) [reply networks/motion/CNN3D/3D_CNN.py]: same here

File: main.py [hunk 28]
```diff
@@ -84,6 +85,13 @@
         # perform splitting
         X_train, y_train, X_test, y_test = ttsplit.fSplitDataset(dAllPatches, dAllLabels, dAllPats, cfg['sSplitting'], patchSize, cfg['patchOverlap'], cfg['dSplitval'], '')
 
+<<<<<<< HEAD
+    if patchSize[0] != 40:
+        X_train, X_test, NewpatchSize = scaling.fscalling(X_train, X_test, patchSize)
```
Comment (thomaskuestner) [inline main.py]: typo: fscalling

File: main.py [hunk 29]
```diff
@@ -94,10 +102,36 @@
...
+    with h5py.File(sDatafile, 'w') as hf:
+         hf.create_dataset('X_train', data=X_train)
+         hf.create_dataset('X_test', data=X_test)
+         hf.create_dataset('y_train', data=y_train)
+         hf.create_dataset('y_test', data=y_test)
+         hf.create_dataset('patchSize', data=patchSize)
+         hf.create_dataset('patchOverlap', data=cfg['patchOverlap'])
+
+# Perform resizing of patches if needed
+X_train, X_test, NewpatchSize, sDatafile = scaling.fscalling(X_train, X_test, patchSize, sDatafile)
```
Comment (thomaskuestner) [inline main.py]: typo: fscalling

File: scaling.py [hunk 30]
```diff
@@ -0,0 +1,106 @@
+
+import h5py
+import numpy as np
+from scipy import interpolate
+import cnn_main
+import time
+
+def fscalling(X_train, X_test, patchSize, sDatafile):
```
Comment (thomaskuestner) [inline scaling.py]: change to fscaling!

File: train_after_resize.py (hunk 13 above)
Comment (thomaskuestner) [reply train_after_resize.py]: combine it with main.py as a separate method (see comments above)

File: networks/motion/CNN2D/motion_head_CNN2D.py (hunk 9 above)
Comment (Sakaaaa) [reply networks/motion/CNN2D/motion_head_CNN2D.py]: Yes, I changed it as keras recommended for compatibility. Later I'll focus on motion_all_CNN2D and I'll abandon this change.

File: scaling.py (hunk 19 above)
Comment (Sakaaaa) [reply scaling.py]: This batch parameter is just for the processing of scaling functions.It just helps the function work faster and has no influence on the result or patches. Can I keep it here?

Comment (thomaskuestner) [reply scaling.py]: in that case you may keep it there

File: main.py [hunk 31]
```diff
@@ -65,27 +68,50 @@
...
-        iLabels = cfg['selectedDatabase']['labelref'] + cfg['selectedDatabase']['labelart']
-        for ipat, pat in enumerate(dbinfo.lPats):
-            if os.path.exists(dbinfo.sPathIn + os.sep + pat + os.sep + dbinfo.sSubDirs[1]):
-                for iseq, seq in enumerate(lDatasets):
-                    # patches and labels of reference/artifact
-                    tmpPatches, tmpLabels  = datapre.fPreprocessData(os.path.join(dbinfo.sPathIn, pat, dbinfo.sSubDirs[1], seq), cfg['patchSize'], cfg['patchOverlap'], 1, cfg['sLabeling'])
-                    dAllPatches = np.concatenate((dAllPatches, tmpPatches), axis=0)
-                    dAllLabels = np.concatenate((dAllLabels, iLabels[iseq]*tmpLabels), axis=0)
-                    dAllPats = np.concatenate((dAllPats, ipat*np.ones((tmpLabels.shape[0],1), dtype=np.int)), axis=0)
+        X_train = []
```
Comment (thomaskuestner) [inline main.py]: This looks good.
@TokenJan @ZiweiWU713 @SebastianMilde Can you confirm that the previous version is still running with these modifications?

**Comment by thomaskuestner** [comment]
@TokenJan @SebastianMilde @Sakaaaa Please check if old version is still running on this new `main.py`

**Comment by TokenJan** [comment]
@thomaskuestner @Sakaaaa it works fine with the existing code but I think `scaling.py` should be put into the util folder and the comment [here](https://github.com/thomaskuestner/CNNArt/pull/29/files#diff-c727d4183a87a0ba438b02a58f40ce3eR83) in the config.yml is confused. Should `sTraingMethod` be set as `PriorScale` instead of  `scalingPrior`? Because in main.py [here](https://github.com/thomaskuestner/CNNArt/pull/29/files#diff-5bc02cefb3ea9e27f1a6776eabd1935dR73) `PriorScale` is used.

**Comment by thomaskuestner** [comment]
Thanks @TokenJan 
Solved with https://github.com/thomaskuestner/CNNArt/commit/62da45e1781d68f7a91c911c0973e893dcd6466a
